spec/lib/session_spec.rb

```diff
@@ -140,6 +141,56 @@
   end
 
   describe "with errors" do
+    context "with refresh_token" do
+      let(:authorization_error) { File.read(File.join("spec", "fixtures", "wsapi", "authorization_error.html")) }
+      let(:refresh_token) { File.read(File.join("spec", "fixtures", "wsapi", "refresh_token.json")) }
+      let(:user_data) { File.read(File.join("spec", "fixtures", "wsapi", "user.json")) }
+
+      before :each do
+        @failing_request = stub_request(:get, wsapi_url_regexp('/User/1'))
+          .with(headers: {'Zsessionid' => "deadbeefdeadbeef"})
+          .to_return(status: 401, body: authorization_error)
+        @successful_request = stub_request(:get, wsapi_url_regexp('/User/1'))
+          .with(headers: { "Zsessionid" => "new_access_token"})
+          .to_return(status: 200, body: user_data)
+        @refresh_token_request = stub_request(:post, Wsapi::AUTH_URL).to_return(status: 200, body: refresh_token)
+      end
+
+      it "requests a new access token with refresh token" do
+        @wsapi.setup_refresh_token("foobar", "thisisasecret", @refresh_token)
+        user = @wsapi.get_user(1)
+        expect(user.name).to eq("Antti Pitkanen")
+        expect(user.email).to eq("apitkanen@rallydev.com")
+
+        expect(@failing_request).to have_been_made
+        expect(@refresh_token_request).to have_been_made
+        expect(@successful_request).to have_been_made
+      end
+
+      it "calls the block given in setup_refresh_token with new tokens" do
+        expect { |b|
+          @wsapi.setup_refresh_token("foobar", "thisisasecret", @refresh_token, &b)
+          @wsapi.get_user(1)
+        }.to yield_with_args({
+          "id_token"=>"???",
+          "refresh_token"=>"new_refresh_token",
+          "expires_in"=>15552000,
+          "token_type"=>"Bearer",
+          "access_token"=>"new_access_token"
+        })
+      end
+    end
+
+    context "without refresh_token" do
```

Comment:
Minor stuff: wouldn't 403 be raised even with refresh_token? The method seems to be 401 as response status.